### PR TITLE
[Proposal] PromptConfiguration for different models

### PR DESCRIPTION
**Pull request by Montagon**
This PR implements a first idea of how we can have different `PromptConfiguration` for each type of model. In this way, we bring more versatility to users as they can configure each model with all the parameters.

### CODE REVIEW

File: gpt4all-kotlin/src/jvmMain/kotlin/com/xebia/functional/gpt4all/GPT4All.kt [hunk 1]
```diff
@@ -29,148 +30,173 @@ import java.nio.file.Path
...
-                val c = b.toChar()
-                channel.trySend(c.toString())
-              }
+            val llModel = LLModel(path)
+
+
+            override suspend fun createCompletion(request: CompletionRequest): CompletionResult =
+                with(request) {
+                    val promptConfig: GPT4AllPromptConfiguration =
+                        request.promptConfiguration as GPT4AllPromptConfiguration
```
Comment (Montagon) [inline gpt4all-kotlin/src/jvmMain/kotlin/com/xebia/functional/gpt4all/GPT4All.kt:87]: I don't know if there is a better way on doing this

File: examples/kotlin/src/main/kotlin/com/xebia/functional/xef/auto/gpt4all/LocalModels.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    println("🤖 GPT4All loaded: $gpt4All")
+
+//    val url = "https://gpt4all.io/models/ggml-gpt4all-j-v1.3-groovy.bin"
+
+    print("\n🤖 Enter the prompt: ")
+    val userInput = readlnOrNull() ?: return
+
+//    val prompt = "###User:\n$userInput\n###Assistant:"
+
+//    ai(model = gpt4All) { //this: GPT4AllAIScope
```
Comment (Montagon) [inline examples/kotlin/src/main/kotlin/com/xebia/functional/xef/auto/gpt4all/LocalModels.kt:36]: This is just an idea of how this could be in the future

**Comment by javipacheco** [comment]
It's important to find a solution to that. This is a good solution and I'm going to suggest another one, may be easier to implement at this point and we can find a better solution for the future when we have more local models working

Another approach could be to use a `Map` where the user can add all of the extra configurations in the model

```kotlin
class PromptConfiguration(
  val maxDeserializationAttempts: Int = 3,
  val user: String = Role.USER.name,
  val temperature: Double = 0.4,
  val numberOfPredictions: Int = 1,
  val docsInContext: Int = 5,
  val memoryLimit: Int = 5,
  val minResponseTokens: Int = 500,
  val streamToStandardOut: Boolean = false,
  val extras: Map<String, Any> = emptyMap(),
)
```

The user should know the `extras`, but this implementation could be a quick solution when we will add a new model with other configurations

Just another suggestion!

**Comment by raulraja** [comment]
This proposal looks good so far 👍

**Pull request closed without merging**